linux/usage/systemd.md

```diff
@@ -1,57 +1,115 @@
 # systemd
 
-In order to have a command or program run when the Pi boots, you can add it as a service. Once this is done, you can start/stop enable/disable from the linux prompt.
+Raspberry Pi OS uses [`systemd`](https://www.freedesktop.org/wiki/Software/systemd/) to manage the running of services, including controlling what starts when Linux boots.
 
-## Creating a service
+## `rc.local` and service dependencies
+Unlike on older versions of Linux, `systemd` starts services based on their dependencies: services are no longer started in a predefined order. This means it is no longer guaranteed that `rc.local` will run after all other services have started. You should therefore not use `rc.local` to start programs at boot: instead, create a `systemd` service, specifying which other services it depends on.
 
-On your Pi, create a .service file for your service, for example:
+## Service definitions
+`Systemd` defines each service in a separate file. Under `systemd` services are a type of 'unit': units are system resources under the control of `systemd`. The basic format of a service definition is as follows:
 
-myscript.service
+```
+[Unit]
+Description=...
+...
+
+[Service]
+...
+
+[Install]
+...
+```
+
+User services should be placed in `/etc/systemd/user`: the filename dictates the service name. For example, a file named `monitoring.service` would define a service named `monitoring`.
+
+### Example - one-time service
+One-time services are used when you wish to run a program which does some work, then exits: this is in contrast to a service which stays running all the time. Consider the following service definition:
+
+```
+[Unit]
+Description=Run my program once at boot time
+Wants=network-online.target
+After=network-online.target
+
+[Service]
+Type=oneshot
+ExecStart=/home/pi/myprogram.sh
+
+[Install]
+WantedBy=multi-user.target graphical.target
```

Comment:
Although it does not hurt, `graphical.target` is never reached without `multi-user.target`, the first pulls in the second: https://www.freedesktop.org/software/systemd/man/systemd.special.html#multi-user.target
Since `multi-user.target` does not imply a "graphical" login (according to the way it's meant to be used), it is pretty commonly the only target used for regular (non-emergency) boot tasks and services. To keep things simple I'd stick with this practically-being convention and use only:
```
WantedBy=multi-user.target
```
For admins which like to handle things via files manually instead of calling `systemctl` with all it's overhead, e.g. in scripts, or if it's broken, using a single install entry also has the benefit that only a single symlink `/etc/systemd/system/multi-user.target.wants/myprogram.service` needs to be created and removed to enable/disable it the way `systemctl enable`/`disable` would do.

Same below in the example service. I see there the `graphical.target` has been added explicitly, so was there any specific reason for that?

Reply:
I can't remember why I put it in originally. My guess would be I was trying to come up with an example that would be as easy as possible for folk to use to get the service definition do what they want. By making the service dependent on graphical.target, it should start later in boot when there is more chance any other services it might depend on are already running.

I think it is probably better to just remove the dependency on graphical.target - I've already added a line telling folk they need to add any services their service depends on as dependencies.

Reply:
> By making the service dependent on graphical.target, it should start later in boot when there is more chance any other services it might depend on are already running.

`WantedBy` entries have no effect on the startup order, same as `Wants=`, `Requires=` do not. Only `Before=` and `After=` have an effect on the order 🙂.

I see now also this misunderstanding in the example:
```
Requires=mpd.target
```
Only defines that the service will not be started if `mpd.target` is not "enabled" or failed to start before. Since it is a target, it most likely has no `[Install]` section, so cannot be enabled but is `WantedBy=mpd.service` or the other way round, mpd.service contains `Wants=mpd.target`.
To be true, at least the official Debian and Raspbian [mpd](https://packages.debian.org/buster/armhf/mpd/filelist) packages do not contain such a target?

However, if one wants to have an MPD client/player started after the MPD service and as well only if MPD service has been started before:
```
Requires=mpd.service
After=mpd.service
```
If there is an mpd.target indeed, then:
```
After=mpd.target
```
and depending on how it's implemented ("active" or "passive" target) additionally, similar to `network-online.target`:
```
Wants=mpd.target
```

Reply:
Thanks for your review Michael - I think I've got it correct now.

Reply:
Jep, LGTM. The two `After=` could be merged to `After=network-online.target mpd.service` but that's cosmetic at best.